ngraph/core/src/op/interpolate.cpp

```diff
@@ -111,143 +117,518 @@ namespace ngraph
-    : Op({image, output_shape})
+                                 const Output<Node>& scales,
+                                 const Output<Node>& axes,
+                                 const op::v4::Interpolate::InterpolateAttrs& attrs)
+    : Op({image, output_shape, scales, axes})
     , m_attrs(attrs)
 {
     constructor_validate_and_infer_types();
 }
 
-bool op::v3::Interpolate::visit_attributes(AttributeVisitor& visitor)
+op::v4::Interpolate::Interpolate(const Output<Node>& image,
+                                 const Output<Node>& output_shape,
+                                 const Output<Node>& scales,
+                                 const op::v4::Interpolate::InterpolateAttrs& attrs)
+    : Op({image, output_shape, scales})
+    , m_attrs(attrs)
+{
+    constructor_validate_and_infer_types();
+}
+
+bool op::v4::Interpolate::visit_attributes(AttributeVisitor& visitor)
 {
     visitor.on_attribute("attrs", m_attrs);
     return true;
 }
 
-void op::v3::Interpolate::validate_and_infer_types()
+static const char* cannot_define_axes =
+    "Cannot define axes of interpolation when there is "
+    "only two inputs and input data have dynamic rank";
+
+std::vector<int64_t> op::v4::Interpolate::get_axes() const
 {
-    NODE_VALIDATION_CHECK(this,
-                          get_input_element_type(1).is_integral_number(),
-                          "output shape must be an integral number.");
-    set_input_is_relevant_to_shape(1);
+    auto inputs = input_values();
+    if (inputs.size() <= 3)
+    {
+        PartialShape input_shape = PartialShape(get_input_partial_shape(0));
+        if (input_shape.rank().is_dynamic())
+        {
+            throw std::invalid_argument(cannot_define_axes);
+        }
+        const auto input_rank = input_shape.rank().get_length();
+        std::vector<int64_t> default_value(input_rank);
+        for (int64_t i = 0; i < input_rank; ++i)
+        {
+            default_value[i] = i;
+        }
+        return default_value;
+    }
 
-    PartialShape output_shape = PartialShape(get_input_partial_shape(0));
-    if (output_shape.rank().is_static())
+    std::vector<int64_t> result;
+    if (auto axes_node = as_type_ptr<op::Constant>(input_value(3).get_node_shared_ptr()))
     {
-        for (auto axis : m_attrs.axes)
+        result = axes_node->cast_vector<int64_t>();
+    }
+
+    return result;
+}
+
+void op::v4::Interpolate::infer_using_scales(PartialShape& output_shape,
+                                             const std::vector<int64_t>& axes,
+                                             const PartialShape& padded_input_shape)
+{
+    auto const_scales = as_type_ptr<op::Constant>(input_value(2).get_node_shared_ptr());
+    if (!const_scales)
+    {
+        return;
+    }
+
+    auto scales = const_scales->cast_vector<float>();
+    size_t i = 0;
+    for (auto axis : axes)
+    {
+        if (padded_input_shape[axis].is_static())
```

Comment:
If the input shape element value is dynamic then we need to set the output shape element dynamic.

Reply:
It is so, because this function is called by `void op::v4::Interpolate::validate_and_infer_types()` preserves dynamic elements (see lines 227, 228, 241-250).